%\VignetteIndexEntry{Bioconductor LaTeX Style}
%\VignettePackage{BiocStyle}
%\VignetteEngine{knitr::knitr}

\documentclass{beamer}

\usetheme{Boadilla}
\setbeamertemplate{footline}
{
  \leavevmode%
  \hbox{%
  \begin{beamercolorbox}[wd=.5\paperwidth,ht=2.25ex,dp=1ex,right,rightskip=1em]{section in head/foot}%
    \usebeamerfont{subsection in head/foot}\hspace*{2ex}\insertshorttitle
  \end{beamercolorbox}%
  \begin{beamercolorbox}[wd=.4\paperwidth,ht=2.25ex,dp=1ex,left,leftskip=1em]{subsection in head/foot}%
    \usebeamerfont{section in head/foot}\insertsectionhead\hspace*{2ex}
  \end{beamercolorbox}%
  \begin{beamercolorbox}[wd=.1\paperwidth,ht=2.25ex,dp=1ex,right]{date in head/foot}%
    \insertframenumber{} / \inserttotalframenumber\hspace*{2ex}
  \end{beamercolorbox}}%
  \vskip0pt%
}

\AtBeginSection[]{
  \begin{frame}<beamer>
    \tableofcontents[currentsection]
  \end{frame}
}

\newcommand{\software}[1]{\textsl{#1}}
\newcommand\R{\software{R}}
%
\newcommand\Bioconductor{\software{Bioconductor}}
\newcommand{\Rpackage}[1]{\textsl{#1}}
\newcommand\Biocpkg[1]{%
  {\href{http://bioconductor.org/packages/#1}%
    {\Rpackage{#1}}}}
\newcommand\CRANpkg[1]{%
  {\href{http://cran.fhcrc.org/web/packages/#1/index.html}%
    {\Rpackage{#1}}}}
\newcommand{\Rcode}[1]{\texttt{#1}}
\newcommand{\Rfunction}[1]{\Rcode{#1}}
\newcommand{\Robject}[1]{\Rcode{#1}}
\newcommand{\Rclass}[1]{\textit{#1}}

\title{Good software: simple, tidy, rich}
\author{Martin Morgan}
\institute[RPCI]{
  Roswell Park Cancer Institute\\
  Buffalo, NY, USA\\
  \href{mailto:martin.morgan@roswellpark.org}{martin.morgan@roswellpark.org}}
\date{20 April 2017}

\begin{document}

\maketitle

\section*{Acknowledgments}

\begin{frame}{\R{} / \Bioconductor{}}
  \url{https://bioconductor.org}\par
  \url{https://support.bioconductor.org}\medskip\par
  \begin{enumerate}
  \item Statistical analysis and comprehension of high-throughput
    genomic data.
  \item Open -- public version control
  \item Reproducibile research -- vignettes, `experiment data'
    packages, release / devel branches, \ldots
  \item Interoperable -- package re-use, data structures, \ldots
  \item Usable -- documentation, support, \ldots
  \end{enumerate}
\end{frame}

\begin{frame}
  \begin{center}
    \includegraphics[height=.8\textheight]{SummarizedExperiment.pdf}
  \end{center}
\end{frame}

\begin{frame}{Simple, tidy, rich: RNA-seq count data}
  \begin{columns}[t]
    \column{.5\textwidth}
    Vocabulary
    \begin{itemize}
    \item Simple: extensive
    \item Tidy: restricted ‘endomorphisms’
    \item Rich: extensive, ‘meaningful’
    \end{itemize}
    Constraints (e.g., probes \& samples)
    \begin{itemize}
    \item Tidy: implicit
    \item Simple, Rich: explicit
    \end{itemize}
    Flexibility
    \begin{itemize}
    \item Simple, tidy: general-purpose
    \item Rich: specialized
    \end{itemize}

    \column{.5\textwidth}
    Programming contract
    \begin{itemize}
    \item Simple, tidy: limited
    \item Rich: strict
    \end{itemize}
    Lessons learned / best practices
    \begin{itemize}
    \item Considerable value in semantically rich structures
    \item Current implementations trade-off user and developer
      convenience
    \item Endomorphism, simple vocabulary, consistent paradigm aid use
    \end{itemize}
  \end{columns}
\end{frame}

\begin{frame}[fragile]{Pretty big data}
  \begin{itemize}
  \item E.g., single-cell RNA-seq, 30,000 genes by 1.3 million samples.
  \item On-disk representation in hdf5.
  \item Convenient in-memory 'matrix' abstraction for subsetting,
    etc.; easy input of manageable subset.
  \item \url{https://github.com/mtmorgan/TENxGenomics}
  \end{itemize}
\begin{verbatim}
> basename(fl)
[1] "1M_neurons_filtered_gene_bc_matrices_h5.h5"
> (tenx <- TENxGenomics(fl))
class: TENxGenomics 
h5path: ./1M_neurons_filtered_gene_bc_matrices_h5.h5 
dim(): 27998 x 1306127 
> tenk <- tenx[, sample(ncol(tenx), 10000)] ## fast
> m <- as.matrix(tenk)                      ## manageable
> se = SummarizedExperiment(list(tenx))     ## rich
\end{verbatim}
\end{frame}

\begin{frame}{Opportunities}
  Programmer analysts
  \begin{itemize}
  \item \url{https://roswellpark.org/careers}
  \item Programmer / Analyst -- \R{} software development \texttt{4924}
  \item Senior Programmer Analyst -- cloud / new-age sys.~admin \texttt{4932}
  \end{itemize}
  \Bioconductor{} annual conference
  \begin{itemize}
  \item Boston, July 26 (D-day) -- 28.
  \item \url{https://bioconductor.org/BioC2017}
  \end{itemize}
\end{frame}

\begin{frame}{Acknowledgments}
  Core team (current \& recent): Yubo Cheng, Valerie Obenchain, Herv\'e
  Pag\`es, Marcel Ramos, Lori Shepherd, Nitesh Turaga.
  \medskip\par
  Technical advisory board: Vincent Carey, Kasper Hansen, Wolfgang
  Huber, Robert Gentleman, Rafael Irizzary, Levi Waldron, Michael
  Lawrence, Sean Davis, Aedin Culhane
  \medskip\par
  Scientific advisory board: Simon Tavare (CRUK), Paul Flicek
  (EMBL/EBI), Simon Urbanek (AT\&T), Vincent Carey (Brigham \&
  Women's), Wolfgang Huber (EBI), Rafael Irizzary (Dana Farber),
  Robert Gentleman (23andMe)
  \medskip\par
  Research reported in this presentation was supported by the National
  Human Genome Research Institute and the National Cancer Institute of
  the National Institutes of Health under award numbers U41HG004059
  and U24CA180996. The content is solely the responsibility of the
  authors and does not necessarily represent the official views of the
  National Institutes of Health.

\end{frame}

\end{document}
